File: app/src/main/java/com/nx/timer/ClickAnimator.java
package com.nx.timer;

import android.view.View;
import android.view.ViewGroup;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

public final class ClickAnimator {

    private ClickAnimator() {}

    public static void applyToAll(View root) {
        if (root == null) return;

        if (root instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) root;
            for (int i = 0; i < group.getChildCount(); i++) {
                applyToAll(group.getChildAt(i));
            }
        }

        boolean isButtonLike = root.isClickable()
                || root instanceof MaterialButton
                || root instanceof FloatingActionButton
                || root instanceof android.widget.Button;

        if (isButtonLike && root.getVisibility() == View.VISIBLE) {
            if (root.getForeground() == null && !(root instanceof ViewGroup)) {
                if (!(root instanceof MaterialButton)
                        && !(root instanceof FloatingActionButton)) {
                    try {
                        android.util.TypedValue outValue = new android.util.TypedValue();
                        root.getContext().getTheme().resolveAttribute(
                                android.R.attr.selectableItemBackgroundBorderless, outValue, true);
                        root.setForeground(root.getContext().getDrawable(outValue.resourceId));
                    } catch (Exception ignored) {}
                }
            }
            addBounceAnim(root);
        }
    }

    private static void addBounceAnim(final View view) {
        if (view.getTag(R.id.tag_anim_applied) != null) return;
        view.setTag(R.id.tag_anim_applied, true);

        view.setPivotX(view.getWidth() / 2f);
        view.setPivotY(view.getHeight() / 2f);
        view.addOnLayoutChangeListener(new BounceLayoutListener());
        view.setOnTouchListener(new BounceTouchListener());
    }
}

File: app/src/main/kotlin/com/nx/timer/ClickAnimator.kt
package com.nx.timer

import android.view.View
import android.view.ViewGroup
import com.google.android.material.button.MaterialButton
import com.google.android.material.floatingactionbutton.FloatingActionButton

object ClickAnimator {

    @JvmStatic
    fun applyToAll(root: View?) {
        if (root == null) return

        if (root is ViewGroup) {
            for (i in 0 until root.childCount) {
                applyToAll(root.getChildAt(i))
            }
        }

        val isButtonLike = root.isClickable
                || root is MaterialButton
                || root is FloatingActionButton
                || root is android.widget.Button

        if (isButtonLike && root.visibility == View.VISIBLE) {
            if (root.foreground == null && root !is ViewGroup) {
                if (root !is MaterialButton && root !is FloatingActionButton) {
                    try {
                        val outValue = android.util.TypedValue()
                        root.context.theme.resolveAttribute(
                            android.R.attr.selectableItemBackgroundBorderless,
                            outValue,
                            true
                        )
                        root.foreground = root.context.getDrawable(outValue.resourceId)
                    } catch (_: Exception) {
                    }
                }
            }
            addBounceAnim(root)
        }
    }

    private fun addBounceAnim(view: View) {
        if (view.getTag(R.id.tag_anim_applied) != null) return
        view.setTag(R.id.tag_anim_applied, true)

        view.pivotX = view.width / 2f
        view.pivotY = view.height / 2f
        view.addOnLayoutChangeListener(BounceLayoutListener())
        view.setOnTouchListener(BounceTouchListener())
    }
}